h2o-clustering/src/main/java/water/clustering/AssistedClusteringEmbeddedConfigProvider.java

```diff
@@ -0,0 +1,79 @@
+package water.clustering;
+
+import org.apache.log4j.Logger;
+import water.clustering.api.AssistedClusteringRestApi;
+import water.clustering.api.GracefulAsyncRunner;
+import water.init.AbstractEmbeddedH2OConfig;
+import water.init.EmbeddedConfigProvider;
+import water.util.Log;
+
+import java.io.IOException;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+/**
+ * Provides {@link AssistedClusteringEmbeddedConfig} as a result of external assist. The resulting
+ * embedded config provides a flatfile assembled by external service, given to H2O. 
+ * This EmbeddedConfig has no timeout and will wait indefinitely until a flatfifle is received.
+ */
+public class AssistedClusteringEmbeddedConfigProvider implements EmbeddedConfigProvider {
```

Comment:
What is the advantage of a "PUSH" design (the node is waiting for being told where are his colleagues) compared to "POLL" design - the node would be asking the operator to get a list of nodes?

The advantage of "POLL":
- not HTTP endpoint needs to be started,
- the implementation on the client-side is simpler - a simple loop with synchronous HTTP rest calls to operators,
- more robust - if there is a network issue, a repeated call will recover from the failure.

Reply:
The contract is still there on the polled server as well - it needs to provide a flatfile of given format and it needs to provide it correctly. This just swaps the role transfers the complexity on the other side - what was client is the server and vice versa.

- No HTTP endpoint has to be started (but it has to - only on the polled flatfile provider's side)
- The implementation on the h2O side is simpler, but the system polled is not.
- The flatfile provider may call H2O repeatedly as well in case off network failure (when it's not polled and it's in the role of a client of course). In fact, it is often the assistant who started H2O and is able to act upon network issues (Especially in K8S, where custom controllers are far more powerful with more rights than a plain pod - typically. But not only in K8S). This, from my POV places the burden on the right side. If a Kubernetes operator doesn't receive any request, it'd have to time out and try to do some guesswork and diagnose the problem on it's own. Doable, but seems like making a straightforward point of failure not easy to spot.

The advantage was exatly that. Less stress on the assisting service, more on H2O. Besides this factor, the following was my motivation:
1. The assistant service knows exactly when the flatfile is ready and pushes the ready state. Polling means additional calls in vain. Event-driven means more effective usually (even though nobody cares about that).
2. The API definition is easy to distribute and the client is easy to generate. Making clients implement a server to be polled is harder for the client. I can even give out a simple `curl` example and people will immediately get the API contract.
3. The more complex part is written once on H2O's side. And the multiple clients are much easier to do.

Yet polling comes with advantages: It is doable with pure Java and no additional dependencies (not even NanoHTTP). So we'd only depend on abstractembeddedconfig from h2o-core.

One more thought:
HTTP server to be polled is done in no time in Rust. Fast, reliable, extremely small, disposable once not needed. Much better, safer and faster done than in Java, with proper implementation of graceful shutdowns without requirements to pull in megabytes of dependencies - [see](https://hyper.rs/guides/server/echo/). I'm saying this because this would **have been** actually easier path for me.

I'm not against the polling solution, just described my motivation. There are two perspectives - one is design. Where to we want to place the burden. Second is effectivity. Making it even-driven is naturally more effective and everything occurs just when it needs to, as described above. On the other side, polling is "easier" on H2O, where the implementation is definitely more complex (but hey, we have it done now ?). I'm advocate of both, but like the push a little bit more. Any thoughts ?

Reply:
Yes, agreed - it is about what side has the complexity. This is a 1-to-many relationship and in my point of view, it is better to keep the "many" side simpler and the "1" side complex from a debuggability point of view. It is also easier to upgrade the service (operator) than to upgrade H2O (which is addressed by having the library not bundled in the jar).

I am fine with either, just wanted to know your reasons. Thank you for explaining.

Reply:
Would be nice to have opinions of @mn-mikke (especially but on exclusively from SW POV) and @honzasterba :wink:

Reply:
dont have a strong preference, personally I find the PUSH method more elegant and easier to work with